Prepare to land COCORecall, refactoring, RaggedTensor inputs, etc

LukeWood opened the pull request:
COCOBase is not feasible due to the way COCOMaP
must be computed.  The good news that comes
with this is that I am able to drastically
simplify the COCORecall implementation, as well
as make it more specialized.

This PR also updates the way constants on the metric
are tracked in order to support the default `merge_state`
method, provides more thorough documentation, and
merges COCOBase and COCORecall's unit tests

fchollet commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+        self,
+        category_ids,
+        iou_thresholds=None,
+        area_range=(0, 1e9**2),
```
Prefer hardcoding the value of `1e9 ** 2`

fchollet commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+    """COCORecall computes the COCO recall metric.
+
+    Args:
+        iou_thresholds: defaults to [0.5:0.05:0.95].
```
Add type information, e.g. `tuple of floats`

fchollet commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+
+    Args:
+        iou_thresholds: defaults to [0.5:0.05:0.95].
+        category_ids: no default, users must provide.
```
Likewise for other args

qlzh727 replied:
Please sort the arg list and match against the arg orders in method signature.

qlzh727 replied:
Could u provide more details about the category_ids? Its unclear to me how it is going to be used.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+import tensorflow.keras.initializers as initializers
 
-from keras_cv.metrics.coco.base import COCOBase
+from keras_cv.metrics.coco import iou as iou_lib
```
For IOU, we might want to move it out of coco name space later, since this is not only applicable for coco metric.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+        )
+        coco_recall.update_state(y_true, y_pred)
+        coco_recall.result()
+        # > 0.0
```
yea, should we give a accurate numerical result?

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+        coco_recall = COCORecall(
+            max_detections=100,
+            category_ids=[1],
+            area_range=(32**2, 64**2),
```
Might want to add this example to area_range docstring, eg it is expecting a list of 2 ints, which are lower and upper bound of size of the bbox.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+        iou_thresholds: defaults to [0.5:0.05:0.95].
+        category_ids: no default, users must provide.
+        area_range: area range to consider bounding boxes in. Defaults to all.
+        max_detections: number of maximum detections a model is allowed to make.
```
Please add the default value here.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+
+        num_images = tf.shape(y_true)[0]
+
+        iou_thresholds = tf.constant(self.iou_thresholds, dtype=tf.float32)
```
I think tf.function will convert python list into constants by itself, you probably can skip this and use len(iou_thresholds) for num_thresholds if needed.

LukeWood replied:
```
E             TypeError: list indices must be integers or slices, not Tensor
```

LukeWood replied:
This is actually a restriction of autograph right now - it does not work well with python lists.  Talked to Dan about this a bit the other day.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
```diff
@@ -12,11 +12,173 @@
+                "sample_weight is not yet supported in keras_cv COCO metrics."
+            )
+
+        if isinstance(y_true, tf.RaggedTensor):
```
We might want to add a TODO here for native ragged input support, since convert ragged tensor to dense tensor could potentially lead to memory issue and cause OOM.

qlzh727 commented on keras_cv/metrics/coco/utils.py:
```diff
@@ -96,3 +96,75 @@ def sort_bboxes(boxes, axis=5):
+    """matches bounding boxes from y_true to boxes in y_pred.
+
+    Args:
+        y_true: bounding box tensor of shape [num_boxes, 4+].
```
I guess `num_boxes` could be different between y_true and y_pred. We might want to make it clear for that.

qlzh727 commented on keras_cv/metrics/coco/utils.py:
```diff
@@ -96,3 +96,75 @@ def sort_bboxes(boxes, axis=5):
+            # TODO(lukewood): update clause to account for gtIg
+            # if m > -1 and gtIg[m] == 0 and gtIg[gind] == 1:
+
+            if not ious[gt_idx, detection_idx] >= iou:
```
I think we should remove the not and use `ious[gt_idx, detection_idx] < iou`:

qlzh727 commented on keras_cv/metrics/coco/utils.py:
```diff
@@ -96,3 +96,75 @@ def sort_bboxes(boxes, axis=5):
+    return pred_matches.stack()
+
+
+def result(self):
```
What is this method?

LukeWood replied:
Sorry - accidentally added.

qlzh727 commented on keras_cv/metrics/coco/utils.py:
```diff
@@ -96,3 +96,75 @@ def sort_bboxes(boxes, axis=5):
     return boxes_sorted_list.stack()
+
+
+def match_boxes(y_true, y_pred, threshold, ious):
```
Do we have a unit test for this?

LukeWood replied:
added

LukeWood commented:
@fchollet @qlzh727 addressed all comments and think this landed on a nicer API than the original as well.  Made `category_ids` => `class_ids` to make it clearer per Keras terminology.  Also made `area_range` default to `None` and do a simple check instead of doing it the way the original implementation did (making the area range huge).

Hope this looks better

qlzh727 commented on examples/metrics/coco/recall_demo.py:
```diff
@@ -0,0 +1,26 @@
+import tensorflow as tf
```
I guess we can skip this demo, since it is included in the docstring, or if it is not, then we should put it in the docstring.

LukeWood replied:
Yeah, I can remove the demo.

LukeWood replied:
Done

LukeWood commented:
@qlzh727 @fchollet let me know if you can find anything else suboptimal with this API.  Internal users are ready to demo, so it would be good to get the feedback loop start once the API is sufficiently polished.

qlzh727 commented on keras_cv/metrics/coco/recall.py:
````diff
@@ -12,11 +12,185 @@
+    padding is available at `keras_cv_.utils.bbox.pad_bbox_batch_to_shape`.
+
+    ```python
+    coco_recall = keras_cv.metrics.coco.COCORecall(
````
From the __init__, I guess the input API is keras_cv.metrics.COCORecall.

LukeWood replied:
Thanks - good catch.

fchollet reviewed (approved):
LGTM
